### PR TITLE
Replace step! with act! and observe

### LINKED ISSUE #12: observe(env)
Should we add `observe(env)` as an optional interface function? See [this explanation](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/issues/68#issuecomment-643592276) by @findmyway why this may be useful in the multi-agent setting. Or should we even make this mandatory and have `reset!` and `step!` return `nothing`?

### LINKED ISSUE #20: Support for multiple players games
How should we support multiple-player games?

I think a good starting principle is that support for multiple-player games should not impact the interface for standard, one-player RL environments. A user that isn't interested about this part should be able to ignore it completely. Therefore, I think we should only propose adding functions and not modify the behavior of existing ones.

Another question is: what kinds of multiple player games do we want to support? I think that supporting two-players, sequential, zero-sum games is a no brainer as most RL algorithms can be trivially extended to support these games. For more complicated games, I think it is open to debate.

**A starting point:**

There should be a `current_player(env)` function but what should it return? An integer between `1` and `num_players(env)`?

### LINKED ISSUE #25: Functions that depend on the environment state vs functions of the environment itself
I think there is an important distinction to be made between functions that depend on the environment's current state (`valid_actions`, `get_state`, `step!`) and functions that only depend on the environment itself (`actions`, `num_players`).

The current API does not make this distinction and I think it is worth debating whether or not this is a good thing.

In AlphaZero.jl, I make this distinction by dispatching either on `AbstractGame` or on `Type{<:AbstractGame}`. Therefore, one would have to write `available_actions(env)` but `num_actions(typeof(env))`.

* One advantage of this approach is that it allows calling functions such as `num_actions` without initializing a new environment. It also makes it clearer what functions depend on the environment's current state and what functions do not.
* A particularly interesting case is the one of the proposed `setstate!(env, state)` function (see https://github.com/JuliaReinforcementLearning/CommonRLInterface.jl/issues/9). I feel a bit uneasy about it because conceptually speaking, the real operation we often want is to initialize a new simulation from a given state. In many cases, the thing we want to write is `env = Env(state)` or `from_state(Env, state)` and not `env = Env(); setstate!(env, state)`.

Now, there are also drawbacks on dispatching on the environment's type:

* Some information about the environment may not be present in its type. For example, an environment may be partially specified by some arguments given to its constructor. There would be a workaround here, which is to encode these arguments in the environment's type using `Val` but I am not sure this is desirable in general.
* In the specific case of `setstate!(env, state)` vs `Env(state)`, there is potentially an argument for the former, which is that it may be faster sometimes to just overwrite the state of an existing simulation session rather than creating a new simulation from scratch.

Any thoughts on this?

### PR DESCRIPTION
**Pull request by zsunberg**
After discussion in #20, I'm now leaning toward having `act!` and `observe` to replace `step!` since it makes significantly more sense in a 2-player environment.

Please weigh in!

If accepted, this will
- close #20 
- close #12 
- close #25 (by adding statements to the docstrings)

### CODE REVIEW

**Comment by zsunberg** [comment]
@jonathan-laurent I can't request a review from you since you are not a JuliaReinforcementLearning member, but I think it would be good for you to weigh in since this was prompted by #20

**Comment by jonathan-laurent** [comment]
@findmyway Could you please add me to JuliaReinforcementLearning?

**Review by findmyway — COMMENTED** [review]
Looks good to me. My only concern is the name of `observe`. 

First, apologize if my comments in https://github.com/JuliaReinforcementLearning/CommonRLInterface.jl/issues/12#issuecomment-645773631 and https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/issues/68#issuecomment-643592276 caused any confusions there. Inspired by all your discussions, I'm experimenting with a wild idea to remove `observe(env)` (at least in single agent environment). So let's forget those discussions temporarily.

From the example here, the `observe` function returns a state. But I guess it is assumed to return an `observation` in POMDP environments? Then how do we get the inner state in such environments? (With a `state(env)`?) And I think @jonathan-laurent mentioned another two functions `set_state!(env, state)` or `Env(state)` somewhere. So I think people with different backgrounds may have a different understanding of `state` or `observation`. We'd better to clearly define these concepts explicitly somewhere in the doc.

**Comment by findmyway** [comment]
> @findmyway Could you please add me to JuliaReinforcementLearning?

It seems you are already in 😉

File: README.md [hunk 1]
```diff
@@ -9,15 +9,25 @@ This package is designed for two reasons:
...
+
+`AbstractEnv` is a base type for all environments.
+
+The interface has four required functions for all `AbstractEnv`s:
 ```julia
-step!(env, a)   # returns an observation, reward, done, and info
-reset!(env)     # returns an observation
+reset!(env)     # returns nothing
 actions(env)    # returns the set of all possible actions for the environment
+act!(env, a)    # returns a reward, done, and info
```
Comment (jonathan-laurent) [inline README.md]: Should we consider only returning the reward or is this for a next PR?

File: src/CommonRLInterface.jl [hunk 2]
```diff
@@ -4,35 +4,105 @@ using MacroTools
...
+This function is a *static property* of the environment; the value it returns should not change based on the state.
+
+---
 
-Return a collection of all of the actions available to the agent in AbstractEnv `env`.
+    actions(env::AbstractZeroSumEnv, i::Integer)
+
+Return a collection of all the actions available to player i.
+
+This function is a *static property* of the environment; the value it returns should not change based on the state.
```
Comment (jonathan-laurent) [inline src/CommonRLInterface.jl:71]: Maybe we should insist that `valid_action_mask` must be defined with respect to `actions(env)` and not `actions(env, player(env))`.

Also, is there a way to make implementing `actions(env)` mandatory but `actions(env, player)` optional?

**Review by jonathan-laurent — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (zsunberg) [reply README.md]: Right - I'll change it depending on the outcome of #20

File: src/CommonRLInterface.jl (hunk 2 above)
Comment (zsunberg) [reply src/CommonRLInterface.jl:71]: Yes, currently `actions(env)` is mandatory, but `actions(env, player)` is optional.

**Comment by zsunberg** [comment]
> We'd better to clearly define these concepts explicitly somewhere in the doc.

Yes. Documentation #24 will be important after we get the rest of the basic questions about the interface answered.

Informally, the definitions are as follows:
`observe` returns all of the information given to the agent on that step to make decisions.
`state` returns the Markov state of the environment. The behavior of two environments with the same static parameters that start from the same state should be statistically identical.

We may want a trait that indicates whether the environment is fully observable, in which case `observe` and `state` return the same information.

**Pull request merged**